internal/cmd/agent/deployer/monitor/updatestatus.go

```diff
@@ -1,56 +1,150 @@
+		return false
+	}
+	if bd.Status.SyncGeneration == nil {
+		return true
+	}
+	return *bd.Status.SyncGeneration != bd.Spec.Options.ForceSyncGeneration
+}
+
+func IsAgent(bd *fleet.BundleDeployment) bool {
+	return strings.HasPrefix(bd.Name, "fleet-agent")
+}
+
+func ShouldUpdateStatus(bd *fleet.BundleDeployment) bool {
+	if bd.Spec.DeploymentID != bd.Status.AppliedDeploymentID {
+		return false
+	}
+
+	// If the bundle failed to install the status should not be updated. Updating
+	// here would remove the condition message that was previously set on it.
+	if condition.Cond(fleet.BundleDeploymentConditionInstalled).IsFalse(bd) {
+		return false
+	}
+
+	return true
+}
+
+func (m *Monitor) UpdateStatus(ctx context.Context, bd *fleet.BundleDeployment, resources *helmdeployer.Resources) (fleet.BundleDeploymentStatus, error) {
+	logger := log.FromContext(ctx).WithName("MonitorBundle")
+
+	// updateResources mutates bd.Status, so copy it first
+	origStatus := *bd.Status.DeepCopy()
+	bd = bd.DeepCopy()
+	err := m.updateFromResources(ctx, bd, resources)
 	if err != nil {
-		return err
+
+		// Returning an error will cause MonitorBundle to requeue in a loop.
+		// When there is no resourceID the error should be on the status. Without
+		// the ID we do not have the information to lookup the resources to
+		// compute the plan and discover the state of resources.
+		if err == helmdeployer.ErrNoResourceID {
+			return origStatus, nil
+		}
+
+		return origStatus, err
 	}
+	status := bd.Status
+
+	readyError := readyError(status)
+	condition.Cond(fleet.BundleDeploymentConditionReady).SetError(&status, "", readyError)
+
+	status.SyncGeneration = &bd.Spec.Options.ForceSyncGeneration
+	if readyError != nil {
+		logger.Info("Status not ready", "error", readyError)
+	}
+
+	removePrivateFields(&status)
+	return status, nil
+}
+
+// removePrivateFields removes fields from the status, which won't be marshalled to JSON.
+// They would however trigger a status update in apply
+func removePrivateFields(s1 *fleet.BundleDeploymentStatus) {
+	for id := range s1.NonReadyStatus {
+		s1.NonReadyStatus[id].Summary.Relationships = nil
+		s1.NonReadyStatus[id].Summary.Attributes = nil
+	}
+}
+
+func readyError(status fleet.BundleDeploymentStatus) error {
+	if status.Ready && status.NonModified {
+		return nil
+	}
+
+	var msg string
+	if !status.Ready {
+		msg = "not ready"
+		if len(status.NonReadyStatus) > 0 {
+			msg = status.NonReadyStatus[0].String()
+		}
+	} else if !status.NonModified {
+		msg = "out of sync"
+		if len(status.ModifiedStatus) > 0 {
+			msg = status.ModifiedStatus[0].String()
+		}
+	}
+
+	return errors.New(msg)
+}
+
+// updateFromResources updates the status with information from the
+// helm release history and an apply dry run.
+func (m *Monitor) updateFromResources(ctx context.Context, bd *fleet.BundleDeployment, resources *helmdeployer.Resources) error {
+	logger := log.FromContext(ctx).WithName("updateBundleDeploymentStatus")
```

Comment:
Are the logger names okay in this file? `updateFromResources()` uses `updateBundleDeploymentStatus`, while `UpdateStatus()` uses `MonitorBundle`.

If the idea is to use something meaningful to the overall process that is happening, maybe we would use `logr.NewContext` to set the logger name at a higher level, so all deeper calls to `log.FromContext(ctx)` will have it already configured?

Reply:
Ah thanks. I renamed the funcs. So far I only used `log.IntoContext` in higher level files. I'll just pass in the logger directly.